File: src/main/java/me/miki/shindo/management/mods/impl/ComboCounterMod.java
package me.miki.shindo.management.mods.impl;

import me.miki.shindo.management.event.EventTarget;
import me.miki.shindo.management.event.impl.EventAttackEntity;
import me.miki.shindo.management.event.impl.EventDamageEntity;
import me.miki.shindo.management.event.impl.EventRender2D;
import me.miki.shindo.management.event.impl.EventTick;
import me.miki.shindo.management.language.TranslateText;
import me.miki.shindo.management.settings.config.Property;
import me.miki.shindo.management.settings.config.PropertyType;
import me.miki.shindo.management.mods.SimpleHUDMod;
import me.miki.shindo.management.nanovg.font.LegacyIcon;

public class ComboCounterMod extends SimpleHUDMod {

    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ICON, category = "Display")
    private boolean iconSetting = true;

    private long hitTime = -1;
    private int combo, possibleTarget;

    public ComboCounterMod() {
        super(TranslateText.COMBO_COUNTER, TranslateText.COMBO_COUNTER_DESCRIPTION);
    }

    @EventTarget
    public void onRender2D(EventRender2D event) {
        this.draw();
    }

    @EventTarget
    public void onTick(EventTick event) {
        if ((System.currentTimeMillis() - hitTime) > 2000) {
            combo = 0;
        }
    }

    @EventTarget
    public void onAttackEntity(EventAttackEntity event) {
        possibleTarget = event.getEntity().getEntityId();
    }

    @EventTarget
    public void onDamageEntity(EventDamageEntity event) {
        if (event.getEntity().getEntityId() == possibleTarget) {
            combo++;
            possibleTarget = -1;
            hitTime = System.currentTimeMillis();
        } else if (event.getEntity() == mc.thePlayer) {
            combo = 0;
        }
    }

    @Override
    public String getText() {
        if (combo == 0) {
            return "No Combo";
        } else {
            return combo + " Combo";
        }
    }

    @Override
    public String getIcon() {
        return iconSetting ? LegacyIcon.BAR_CHERT : null;
    }
}

File: src/main/kotlin/me/miki/shindo/management/mods/impl/ComboCounterMod.kt
package me.miki.shindo.management.mods.impl

import me.miki.shindo.management.event.EventTarget
import me.miki.shindo.management.event.impl.EventAttackEntity
import me.miki.shindo.management.event.impl.EventDamageEntity
import me.miki.shindo.management.event.impl.EventNVG
import me.miki.shindo.management.event.impl.EventTick
import me.miki.shindo.management.language.TranslateText
import me.miki.shindo.management.mods.SimpleHUDMod
import me.miki.shindo.management.nanovg.font.Lucide
import me.miki.shindo.management.nanovg.font.Shinconic
import me.miki.shindo.management.settings.config.Property
import me.miki.shindo.management.settings.config.PropertyType

class ComboCounterMod : SimpleHUDMod(TranslateText.COMBO_COUNTER, TranslateText.COMBO_COUNTER_DESCRIPTION, Shinconic.MOD_COMBO_COUNTER) {
    @Property(type = PropertyType.BOOLEAN, translate = TranslateText.ICON)
    private val iconSetting = true

    private var hitTime: Long = -1
    private var combo = 0
    private var possibleTarget = 0

    @EventTarget
    fun onRender2D(event: EventNVG?) {
        this.draw()
    }

    @EventTarget
    fun onTick(event: EventTick?) {
        if ((System.currentTimeMillis() - hitTime) > 2000) {
            combo = 0
        }
    }

    @EventTarget
    fun onAttackEntity(event: EventAttackEntity) {
        possibleTarget = event.getEntity().entityId
    }

    @EventTarget
    fun onDamageEntity(event: EventDamageEntity) {
        if (event.getEntity().entityId == possibleTarget) {
            combo++
            possibleTarget = -1
            hitTime = System.currentTimeMillis()
        } else if (event.getEntity() === mc.thePlayer) {
            combo = 0
        }
    }

    override fun getText(): String =
        if (combo == 0) {
            "No Combo"
        } else {
            "$combo Combo"
        }

    override fun getIcon(): String? = if (iconSetting) Lucide.BAR_CHART else null
}
